spring-xd-dirt/src/main/java/org/springframework/xd/dirt/stream/XDParser.java

```diff
@@ -43,4 +45,29 @@
 	 *         described by the DSL.
 	 */
 	List<ModuleDescriptor> parse(String name, String config, ParsingContext type);
+
+	public static class DefinitionOrException {
```

Comment:
Agree 100% with your thoughts on where the "inner" parser should go.

It sounds like the best compromise for now would be to take the decoration path for the multi-stream "dry run" parser, and while at it maybe change the return value from a List of instances of classes that include "OR" in the name to something more like a MultiStreamParseResult or whatever.